src/spec.c

```diff
@@ -1188,41 +1191,59 @@ static void IndexStats_RdbSave(RedisModuleIO *rdb, IndexStats *stats) {
 
 static threadpool reindexPool = NULL;
 
-typedef struct IndexesScanner {
-  const char *spec_name_opt;  // can be null
-  IndexSpec *spec_opt;
-  size_t scannedKeys, totalKeys;
-} IndexesScanner;
-
-static IndexesScanner *IndexesScanner_New(IndexSpec *spec_opt) {
+static IndexesScanner *IndexesScanner_New(IndexSpec *spec) {
+  if (!spec && global_spec_scanner) {
+    return NULL;
+  }
   IndexesScanner *scanner = rm_calloc(1, sizeof(IndexesScanner));
-  scanner->spec_name_opt = spec_opt ? rm_strdup(spec_opt->name) : NULL;
+  scanner->global = !spec;
+  scanner->spec = spec;
   scanner->scannedKeys = 0;
+  scanner->cancelled = false;
   RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
   scanner->totalKeys = RedisModule_DbSize(ctx);
   RedisModule_FreeThreadSafeContext(ctx);
-  if (spec_opt) {
-    __sync_fetch_and_add(&spec_opt->pending_indexing_ops, 1);
-    spec_opt->keysIndexed = 0;
-    spec_opt->keysTotal = scanner->totalKeys;
+
+  if (spec) {
+    // scan already in progress?
+    if (spec->scanner) {
+      // cancel ongoing scan, keep on_progress indicator on
+      IndexesScanner_Cancel(spec->scanner, true);
+    }
+    spec->scanner = scanner;
+    spec->scan_in_progress = true;
   } else {
-    __sync_fetch_and_add(&pending_global_indexing_ops, 1);
+    global_spec_scanner = scanner;
   }
+
   return scanner;
 }
 
 void IndexesScanner_Free(IndexesScanner *scanner) {
-  if (scanner->spec_name_opt) {
-    rm_free((void *)scanner->spec_name_opt);
-    if (scanner->spec_opt) {
-      __sync_fetch_and_sub(&scanner->spec_opt->pending_indexing_ops, 1);
+  if (global_spec_scanner == scanner) {
+    global_spec_scanner = NULL;
+  } else if (!scanner->cancelled) {
+    if (scanner->spec && scanner->spec->scanner == scanner) {
+      scanner->spec->scanner = NULL;
+      scanner->spec->scan_in_progress = false;
     }
-  } else {
-    __sync_fetch_and_sub(&pending_global_indexing_ops, 1);
   }
+
   rm_free(scanner);
 }
 
+void IndexesScanner_Cancel(IndexesScanner *scanner, bool still_in_progress) {
+  if (scanner->cancelled) {
+    return;
+  }
+  if (!scanner->global && scanner->spec) {
```

Comment:
Could it not be global but not have a spec? I think you have duplicate information here. Also do we even cancel a global scan? If we don't we can assert the scanner is not global.

Reply:
It certainly seems like `scanner->global` and `!scanner->spec` signal the same thing.
The point here is during cancellation, `scanner->spec` will become NULL (this could happen on a spec removal, where we must set `scanner->spec` to NULL to avoid referencing invalid memory).
So I preferred to have some sort of possible duplication in favor of cleared indication of global-ness.
The whole handing of the scanner and indexes is pretty delicate and I hope that we'll have better methodology of handling async operations (like better sense of object ownership).